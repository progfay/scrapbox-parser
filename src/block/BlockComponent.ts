export type BlockComponentType = {
  readonly indent: number
  readonly text: string
}

export const convertToBlockComponent = (block: string): BlockComponentType => ({
  indent: block.match(/^\s+/)?.[0].length ?? 0,
  text: block
})

export const convertToBlockComponents = (blocks: string): BlockComponentType[] => (
  blocks.split('\n').map(convertToBlockComponent)
)
