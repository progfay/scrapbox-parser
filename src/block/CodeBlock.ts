import type { Row } from "./Row";

export interface CodeBlockPack {
	type: "codeBlock";
	rows: Row[];
}

/**
 * Scrapbox {@link https://scrapbox.io/help/Syntax#58348ae2651ee500008d67df | code block} type
 */
export interface CodeBlock {
	indent: number;
	type: "codeBlock";
	fileName: string;
	content: string;
}

export const convertToCodeBlock = (pack: CodeBlockPack): CodeBlock => {
	const {
		rows: [head, ...body],
	} = pack;
	const { indent = 0, text = "" } = head ?? {};
	const fileName: string = text.replace(/^\s*code:/, "");

	return {
		indent,
		type: "codeBlock",
		fileName,
		content: body
			.map((row: Row): string => row.text.substring(indent + 1))
			.join("\n"),
	};
};
