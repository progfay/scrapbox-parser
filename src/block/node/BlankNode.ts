import { createNodeParser } from "./creator";
import { createPlainNode } from "./PlainNode";

import type { NodeParser } from "./index";
import type { BlankNode, PlainNode } from "./type";
import type { NodeCreator } from "./creator";

const blankRegExp = /\[\s+\]/;

const createBlankNode: NodeCreator<BlankNode | PlainNode> = (raw, opts) =>
  opts.context === "table"
    ? createPlainNode(raw, opts)
    : [
        {
          type: "blank",
          raw,
          text: raw.substring(1, raw.length - 1),
        },
      ];

export const BlankNodeParser: NodeParser = createNodeParser(createBlankNode, {
  parseOnNested: false,
  parseOnQuoted: true,
  patterns: [blankRegExp],
});
