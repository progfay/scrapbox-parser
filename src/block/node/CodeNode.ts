import { createPlainNode } from "./PlainNode";
import { createNodeParser } from "./creator";

import type { NodeCreator } from "./creator";
import type { NodeParser } from "./index";
import type { CodeNode, PlainNode } from "./type";

const codeRegExp = /`.*?`/;

const createCodeNode: NodeCreator<CodeNode | PlainNode> = (raw, opts) =>
	opts.context === "table"
		? createPlainNode(raw, opts)
		: [
				{
					type: "code",
					raw,
					text: raw.substring(1, raw.length - 1),
				},
			];

export const CodeNodeParser: NodeParser = createNodeParser(createCodeNode, {
	parseOnNested: false,
	parseOnQuoted: true,
	patterns: [codeRegExp],
});
