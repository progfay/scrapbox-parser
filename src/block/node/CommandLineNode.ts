import { createPlainNode } from "./PlainNode.ts";
import { createNodeParser } from "./creator.ts";

import type { NodeCreator } from "./creator.ts";
import type { NodeParser } from "./index.ts";
import type { CommandLineNode, PlainNode } from "./type.ts";

const commandLineRegExp = /^[$%] .+$/;

const createCommandLineNode: NodeCreator<CommandLineNode | PlainNode> = (
	raw: string,
	opts,
) => {
	if (opts.context === "table") {
		return createPlainNode(raw, opts);
	}

	const symbol = raw[0] ?? "";
	const text = raw.substring(2);

	return [
		{
			type: "commandLine",
			raw,
			symbol,
			text,
		},
	];
};

export const CommandLineNodeParser: NodeParser = createNodeParser(
	createCommandLineNode,
	{
		parseOnNested: false,
		parseOnQuoted: false,
		patterns: [commandLineRegExp],
	},
);
