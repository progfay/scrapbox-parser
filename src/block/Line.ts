import { BlockComponentType } from './BlockComponent'
import { LineNodeType, convertToLineNodes } from './node'

export type LineComponentType = {
  readonly type: 'line'
  readonly component: BlockComponentType
}

export type LineType = {
  readonly indent: number
  readonly type: 'line'
  readonly nodes: ReadonlyArray<LineNodeType>
}

export const convertToLine = (lineComponent: LineComponentType): LineType => {
  const { indent, text } = lineComponent.component
  return {
    indent,
    type: 'line',
    nodes: convertToLineNodes(text.substring(indent))
  }
}
