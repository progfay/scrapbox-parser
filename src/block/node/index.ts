import { QuoteNodeParser } from './QuoteNode'
import { HelpfeelNodeParser } from './HelpfeelNode'
import { StrongImageNodeParser } from './StrongImageNode'
import { StrongIconNodeParser } from './StrongIconNode'
import { StrongNodeParser } from './StrongNode'
import { FormulaNodeParser } from './FormulaNode'
import { DecorationNodeParser } from './DecorationNode'
import { CodeNodeParser } from './CodeNode'
import { CommandLineNodeParser } from './CommandLineNode'
import { BlankNodeParser } from './BlankNode'
import { ImageNodeParser } from './ImageNode'
import { ExternalLinkNodeParser } from './ExternalLinkNode'
import { GoogleMapNodeParser } from './GoogleMapNode'
import { InternalLinkNodeParser } from './InternalLinkNode'
import { IconNodeParser } from './IconNode'
import { HashTagNodeParser } from './HashTagNode'
import { PlainNodeParser } from './PlainNode'

import type { Node } from './type'

export interface NodeParserOption {
  nested: boolean
  quoted: boolean
}
export type NextNodeParser = () => Node[]
export type NodeParser = (text: string, opts: NodeParserOption, next?: NextNodeParser) => Node[]

const FalsyEliminator: NodeParser = (text, _, next) => {
  if (text === '') return []
  return next?.() ?? []
}

const combineNodeParsers = (...parsers: NodeParser[]) => (
  text: string = '',
  opts: NodeParserOption = { nested: false, quoted: false }
): Node[] =>
  parsers.reduceRight(
    (acc: NextNodeParser, parser: NodeParser): NextNodeParser => () => parser(text, opts, acc),
    () => PlainNodeParser(text, opts)
  )()

export const convertToNodes = combineNodeParsers(
  FalsyEliminator,
  QuoteNodeParser,
  HelpfeelNodeParser,
  CodeNodeParser,
  CommandLineNodeParser,
  FormulaNodeParser,
  BlankNodeParser,
  DecorationNodeParser,
  StrongImageNodeParser,
  StrongIconNodeParser,
  StrongNodeParser,
  ImageNodeParser,
  ExternalLinkNodeParser,
  IconNodeParser,
  GoogleMapNodeParser,
  InternalLinkNodeParser,
  HashTagNodeParser
)
