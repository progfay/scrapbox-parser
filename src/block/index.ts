import { convertToCodeBlock } from "./CodeBlock";
import { convertToLine } from "./Line";
import { convertToTable } from "./Table";
import { convertToTitle } from "./Title";

import type { CodeBlock } from "./CodeBlock";
import type { Line } from "./Line";
import type { Pack } from "./Pack";
import type { Table } from "./Table";
import type { Title } from "./Title";

/**
 * Scrapbox block type
 */
export type Block = Title | CodeBlock | Table | Line;

export const convertToBlock = (pack: Pack): Block => {
	switch (pack.type) {
		case "title":
			return convertToTitle(pack);

		case "codeBlock":
			return convertToCodeBlock(pack);

		case "table":
			return convertToTable(pack);

		case "line":
			return convertToLine(pack);
	}
};
