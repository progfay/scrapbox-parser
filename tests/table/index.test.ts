/* global describe it expect */
/* eslint-disable no-tabs, no-irregular-whitespace */

import '../jest-setup'
import { LineNodeType } from '../../src/block/node'

const defaultTableCells: LineNodeType[][][] = [
  [
    [
      {
        type: 'plain',
        text: '1'
      }
    ],
    [
      {
        type: 'plain',
        text: '2'
      }
    ],
    [
      {
        type: 'plain',
        text: '3'
      }
    ]
  ],
  [
    [{
      type: 'plain',
      text: '1 '
    }],
    [
      {
        type: 'plain',
        text: '2 '
      }
    ],
    [
      {
        type: 'plain',
        text: '3'
      }
    ]
  ],
  [
    [
      {
        type: 'plain',
        text: '------'
      }
    ],
    [
      {
        type: 'plain',
        text: '------'
      }
    ],
    [
      {
        type: 'plain',
        text: '------'
      }
    ]
  ],
  [
    [
      {
        type: 'plain',
        text: 'a'
      }
    ],
    [
      {
        type: 'plain',
        text: 'b'
      }
    ],
    [
      {
        type: 'plain',
        text: 'c'
      }
    ]
  ]
]

describe('Table', () => {
  it('Simple table', () => {
    expect(`table:hello
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1 ${'\t'}2 ${'\t'}3
${'\t'}------${'\t'}------${'\t'}------
${'\t'}a${'\t'}b${'\t'}c`).toEqualWhenParsing([
      {
        indent: 0,
        type: 'table',
        fileName: 'hello',
        cells: defaultTableCells
      }
    ])
  })

  it('Bulleted table', () => {
    expect(` table:bulleted
 ${'\t'}1${'\t'}2${'\t'}3
 ${'\t'}1 ${'\t'}2 ${'\t'}3
 ${'\t'}------${'\t'}------${'\t'}------
 ${'\t'}a${'\t'}b${'\t'}c`).toEqualWhenParsing([
      {
        indent: 1,
        type: 'table',
        fileName: 'bulleted',
        cells: defaultTableCells
      }
    ])
  })

  it('Table with empty cells', () => {
    expect(`table:${' '}
${'\t'} ${'\t'}　${'\t'}${'  '}
${'\t'}${'\t'}${'\t'}`).toEqualWhenParsing([
      {
        indent: 0,
        type: 'table',
        fileName: ' ',
        cells: [
          [
            [
              {
                type: 'plain',
                text: ' '
              }
            ],
            [
              {
                type: 'plain',
                text: '　'
              }
            ],
            [
              {
                type: 'plain',
                text: '  '
              }
            ]
          ],
          [
            [],
            [],
            []
          ]
        ]
      }
    ])
  })

  it('Staggered table', () => {
    expect(`table:Staggered
${'\t'}1${'\t'}2${'\t'}3${'\t'}4
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1
${'\t'}1${'\t'}2
${'\t'}`).toEqualWhenParsing([
      {
        indent: 0,
        type: 'table',
        fileName: 'Staggered',
        cells: [
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ],
            [
              {
                type: 'plain',
                text: '2'
              }
            ],
            [
              {
                type: 'plain',
                text: '3'
              }
            ],
            [
              {
                type: 'plain',
                text: '4'
              }
            ]
          ],
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ],
            [
              {
                type: 'plain',
                text: '2'
              }
            ],
            [
              {
                type: 'plain',
                text: '3'
              }
            ]
          ],
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ]
          ],
          [
            [
              {
                type: 'plain',
                text: '1'
              }
            ],
            [
              {
                type: 'plain',
                text: '2'
              }
            ]
          ],
          [
            []
          ]
        ]
      }
    ])
  })

  it('Consecutive table', () => {
    expect(`table:hello
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1 ${'\t'}2 ${'\t'}3
${'\t'}------${'\t'}------${'\t'}------
${'\t'}a${'\t'}b${'\t'}c
table:hello
${'\t'}1${'\t'}2${'\t'}3
${'\t'}1 ${'\t'}2 ${'\t'}3
${'\t'}------${'\t'}------${'\t'}------
${'\t'}a${'\t'}b${'\t'}c`).toEqualWhenParsing([
      {
        indent: 0,
        type: 'table',
        fileName: 'hello',
        cells: defaultTableCells
      },
      {
        indent: 0,
        type: 'table',
        fileName: 'hello',
        cells: defaultTableCells
      }
    ])
  })

  it('Table with link', () => {
    expect(`table:table with link
${'\t'}[Link]${'\t'}This is [Link]`).toEqualWhenParsing([
      {
        indent: 0,
        type: 'table',
        fileName: 'table with link',
        cells: [
          [
            [
              {
                type: 'link',
                pathType: 'relative',
                href: 'Link',
                content: ''
              }
            ],
            [
              {
                type: 'plain',
                text: 'This is '
              },
              {
                type: 'link',
                pathType: 'relative',
                href: 'Link',
                content: ''
              }
            ]
          ]
        ]
      }
    ])
  })
})
