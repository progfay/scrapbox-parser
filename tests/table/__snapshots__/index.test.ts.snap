// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Bulleted table 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1 ",
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2 ",
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "a",
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "b",
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "c",
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "bulleted",
    "indent": 1,
    "type": "table",
  },
]
`;

exports[`Table Consecutive table 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1 ",
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2 ",
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "a",
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "b",
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "c",
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1 ",
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2 ",
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "a",
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "b",
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "c",
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Node in table cells 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": "#hashtag",
            "text": "#hashtag",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[* deco]",
            "text": "[* deco]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[ ]",
            "text": "[ ]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "\`code\`",
            "text": "\`code\`",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "https://external.com",
            "text": "https://external.com",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[https://external.com]",
            "text": "[https://external.com]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[left https://external.com]",
            "text": "[left https://external.com]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[https://external.com right]",
            "text": "[https://external.com right]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[$ x]",
            "text": "[$ x]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[N35.6812362,E139.7649361]",
            "text": "[N35.6812362,E139.7649361]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "#hashTag",
            "text": "#hashTag",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "? helpfeel",
            "text": "? helpfeel",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "$ commandLine",
            "text": "$ commandLine",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "path": "progfay",
            "pathType": "relative",
            "raw": "[progfay.icon]",
            "type": "icon",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[https://image.com/image.png]",
            "text": "[https://image.com/image.png]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "content": "",
            "href": "link",
            "pathType": "relative",
            "raw": "[link]",
            "type": "link",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "plain",
            "text": "plain",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "> quote",
            "text": "> quote",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[[progfay.icon]]",
            "text": "[[progfay.icon]]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[[https://image.com/image.png]]",
            "text": "[[https://image.com/image.png]]",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "[[strong]]",
            "text": "[[strong]]",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "node in table cells",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Simple table 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1 ",
            "text": "1 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2 ",
            "text": "2 ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "a",
            "text": "a",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "b",
            "text": "b",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "c",
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Staggered table 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "4",
            "text": "4",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [
          Object {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [],
      ],
    ],
    "fileName": "Staggered",
    "indent": 0,
    "type": "table",
  },
]
`;

exports[`Table Table with empty cells 1`] = `
Array [
  Object {
    "cells": Array [
      Array [
        Array [
          Object {
            "raw": " ",
            "text": " ",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "　",
            "text": "　",
            "type": "plain",
          },
        ],
        Array [
          Object {
            "raw": "  ",
            "text": "  ",
            "type": "plain",
          },
        ],
      ],
      Array [
        Array [],
        Array [],
        Array [],
      ],
    ],
    "fileName": " ",
    "indent": 0,
    "type": "table",
  },
]
`;
